File: palindrome.js
// Given a string, determine if it is a palindrome, considering only alphanumeric characters and ignoring cases.


// Write without regex

var isPalindrome = function(s) {
  s = s.toLowerCase().split('');
  head = 0;
  tail = s.length - 1;

  while (head < tail) {
    while (head < tail && /\W/.test(s[head])) {
      head++;
    }
    while (head < tail && /\W/.test(s[tail])) {
      tail--;
    }

    if (s[head] !== s[tail]) {
      return false;
    }

    head++;
    tail--;
  }

  return true;
};

File: palindrome.rb
# Given a string, determine if it is a palindrome, considering only alphanumeric characters and ignoring cases.

def is_palindrome(s)
  s = s.gsub(/\W/, '').downcase.chars
  head = 0
  tail = s.length - 1

  while (head < tail) do
    return false unless s[head] == s[tail]

    head += 1
    tail -= 1
  end

  true
end
